[Merged by Bors] - feat: add CI to auto-label PRs

joneugster opened the pull request:
Add command `lake exe autolabel 17062` which uses `git diff` to find changes compared to `master`, searches for applicable GitHub topic labels (based on a defined mapping), and adds them to the PR using `gh`.

Add GitHub workflow calling the script on creation of a new PR, automatically adding topic labels to the PR if applicable.

---

- [x] Counterproposal to #15938 (only one of the two should be merged!)




Currently two potential issues come to mind:

- anybody calling `lake exe autolabel 10000` locally on their machine might add random labels to an arbitrary PR. This will say their name (and not a bot). I guess this is fine since it is by choice that any contributor can arbitrarily modify github labels...

See PR in action here: #17088 and #17083 (no label as two were found, see [CI run](https://github.com/leanprover-community/mathlib4/actions/runs/11013546073/job/30582139194#step:4:46))

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> Mathlib.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

joneugster commented:
@adomani what's your opinion? Would you have the time to check that the logic is equivalent to yours? (I think it should be...)

adomani commented on scripts/autolabel.lean:
```diff
@@ -0,0 +1,160 @@
+/--
+Mathlib Labels and their corresponding folders. Add new labels and folders here!
+-/
+private def mathlibLabels : Std.HashSet Label := Std.HashSet.ofList [
```
```suggestion
def mathlibLabels : Array Label := #[
```
Here is my reasoning for the changes:
* a file in `scripts` will not be imported, so `private` seems superfluous;
* `mathlibLabels` is used only once and immediately converted to an `Array`, so it might as well be an `Array`!

adomani commented on scripts/autolabel.lean:
```diff
@@ -0,0 +1,160 @@
+    dirs := #[
+      ".github" / "workflows" ]}]
+
+def getMatchingLabels (files : Array String) : Array Label :=
```
Would you mind adding a doc-string to this definition?  It shows that I've been playing with linters for a while, doesn't it?  :smile:

adomani commented on scripts/autolabel.lean:
```diff
@@ -0,0 +1,160 @@
+      "Mathlib" /" RingTheory",
+      "Mathlib" / "GroupTheory",
+      "Mathlib" / "RepresentationTheory",
+      "Mathlib" / "LinearAlgebra" ]},
```
```suggestion
      "Mathlib" / "LinearAlgebra" ] },
```
just for consistency with the space at the opening of the bracket.

joneugster replied:
I don't have a strong opinion on this, so I replaced the Array with the result of pretty-printing it (modulo some line breaks)

adomani commented on scripts/autolabel.lean:
```diff
@@ -0,0 +1,160 @@
+        args := #["pr", "edit", n, "--add-label", label] }
+      println s!"Added label: {label}"
+    | none =>
+      println s!"No PR-number provided, skipping adding labels.
```
```suggestion
      println s!"No PR-number provided, skipping adding labels. \
```
maybe?

adomani reviewed (commented):
Thanks a lot for implementing an alternative!

I like the executable!

I find adding more labels a little less streamlined than in the other PR, but I also don't think that adding new labels is something that will happen a lot!

Would you mind adding some tests for the `getMatchingLabels` function?

grunweg reviewed (commented):
Thanks for writing this alternative. Because it avoids environment extensions, I can review it in depth - the implementation looks good to me. (I have one question about correctness, and otherwise just documentation/style comments.)

grunweg commented:
My personal opinion on which of these is preferrable:
- the other PR makes it much more convenient to add new labels, to the point of hiding all core logic behind an environment extension - but that means 100-200 lines for environment extensions, make that implementation more complicated and harder to understand. (I certainly don't understand that part, at the moment.)
- this PR just does the "minimal" thing, in a nice way. Adding a new label requires editing the list of labels in the implementation, but that seems reasonable enough (and the implementation is not too hard to read). So, I think it's good enough in practice.

grunweg commented:
Two more questions about this PR:
- should the executable error if not automatic label was found (e.g., when a directory `Geometry/Hyperbolic` is added, which is not covered by any rule)? (You'd have to exclude `scripts` from that error, but otherwise it looks like it could work.)
- if a PR is labelled wrongly, can this be corrected manually? Or does the workflow run on every *push* (as opposed to once, when the PR is filed)

In any case: thanks a lot, both of you, for working on this - this piece of automation will be great to have!

joneugster commented on scripts/autolabel.lean:
```diff
@@ -0,0 +1,150 @@
+/--
+Mathlib Labels and their corresponding folders. Add new labels and folders here!
+-/
+def mathlibLabels : Array Label := #[
+  { label := "t-algebra",
+    dirs := #[
+      "Mathlib" / "FieldTheory",
+      "Mathlib" / " RingTheory",
+      "Mathlib" / "GroupTheory",
+      "Mathlib" / "RepresentationTheory",
+      "Mathlib" / "LinearAlgebra"] },
```
@grunweg re hiding implementation from the user, this could be a separate JSON file instead of a `def` if you'd prefer that?

adomani replied:
We could also use an `IO.Ref` and recycle some of the syntax from the other PR, but avoiding Environment extensions.  After all, everything happens in one file only.

joneugster replied:
nvm, that would be less nice with the filepaths...

grunweg replied:
I think the current logic is good enough.

joneugster commented on scripts/autolabel.lean:
```diff
@@ -0,0 +1,160 @@
+  { label := "t-algebraic-geometry",
+    dirs := #[
+      "Mathlib" / "AlgebraicGeometry",
+      "Mathlib" / "Geometry.RingedSpace" ]},
```
Dumb question ( also @adomani ), how do I test a script? I don't think I can add `test/autolabel.lean` and `import scripts.autolabel`, can I?

adomani replied:
I had in mind something like
```lean
run_cmd
  if getMatchingLabels input != desiredOutput then
    logWarning "Test failed"
```
Not too highbrow, but at least it shows what some representative inputs get converted to.

joneugster replied:
ok, I've did the same with `#guard` which is about equivalent

joneugster commented on scripts/autolabel.lean:
```diff
@@ -0,0 +1,160 @@
+
+/-- `args` is expected to have length 1, and the first argument is the PR number. -/
+unsafe def main (args : List String): IO Unit := do
+  let prNumber? := args[0]?
```
Do you mean as `println` as pushed now? or is there a better way?

grunweg replied:
The current way looks good to me.

joneugster commented on scripts/autolabel.lean:
```diff
@@ -0,0 +1,160 @@
+  mathlibLabels.toArray.filter fun label =>
+    -- modified files which are not excluded by the label
+    let notExcludedFiles := files.filter fun file =>
+      label.exclusions.map (!·.toString.isPrefixOf file) |>.all (·)
```
Good spot, the old version had a workaround I didn't copy!

joneugster commented on .github/workflows/add_label_from_diff.yaml:
```diff
@@ -0,0 +1,39 @@
+name: Autolabel PRs
+
+on:
+  pull_request:
+    types: [opened]
+  workflow_dispatch:
```
@grunweg I updated the workflow, so now it only triggers once when the PR is opened, but there is also the option to dispatch it manually from the github UI

joneugster replied:
There is one catch with this:

Since the workflow only triggers once, one could add a new folder in say the second commit to the PR. This might lead to a passing CI but a new PR created will then complain that not all of Mathlib is covered by the labels....

bryangingechen replied:
Do we need a test that all of mathlib is covered by labels? If so, could we run it say just once a week or something? I'd really like to avoid forcing people to have to solve unrelated CI issues in their PRs if something sneaks through in someone else's PR...

joneugster replied:
I ended up that this only triggers a warning. So there might be an unexpected warning on future PRs, but at least it should not cause anything to fail.

grunweg commented on scripts/autolabel.lean:
```diff
@@ -0,0 +1,195 @@
+This file contains the script to automatically assign a GitHub label to a PR.
+
+`AutoLabel.mathlibLabels` contains an assignment of github labels to folders inside
+the mathlib repository. The sctript uses `git diff` to determine which files
```
```suggestion
the mathlib repository. The script uses `git diff` to determine which files

grunweg commented on scripts/autolabel.lean:
```diff
@@ -0,0 +1,195 @@
+      #["Mathlib" / ("".intercalate (label.splitOn "-" |>.drop 1 |>.map .capitalize))]
+    else #[]
+  /-- Array of paths which should be excluded.
+  Any modifications to a file in an excluded path is ignored for the purposes of labelling. -/
```
```suggestion
  Any modifications to a file in an excluded path are ignored for the purposes of labelling. -/

grunweg commented on scripts/autolabel.lean:
```diff
@@ -0,0 +1,195 @@
+  deriving BEq, Hashable
+
+/--
+Mathlib Labels and their corresponding folders. Add new labels and folders here!
```
```suggestion
Mathlib labels and their corresponding folders. Add new labels and folders here!

grunweg commented on scripts/autolabel.lean:
```diff
@@ -0,0 +1,195 @@
+  (dir / "").normalize.toString.isPrefixOf path.normalize.toString
+
+/--
+Return all labels names for labels in `mathlibLabels` which are matching
```
How about this?
```suggestion
Return all names of labels in `mathlibLabels` which match

grunweg commented on scripts/autolabel.lean:
```diff
@@ -0,0 +1,195 @@
+    IO.Process.exit 1
+  let prNumber? := args[0]?
+
+  -- validate that all paths in `mathilbLabels` actually exist
```
```suggestion
  -- validate that all paths in `mathlibLabels` actually exist

grunweg commented on scripts/autolabel.lean:
```diff
@@ -0,0 +1,195 @@
+
+  match labels with
+  | #[] =>
+    println s!"No applicable labels found!"
```
I like the validation check you added. I meant to (also) throw an error if this branch is reached --- unless the PR modifies e.g. only `scripts`.

joneugster replied:
I'm a bit hesitant to make this failing for the following reasons:

- it will mean the PR gets a red cross because there was no applicable label, even if the PR is passing CI. (afaik there is no way to make CI pass despite a single workflow failing).
- There are still plenty of folders/files which are not covered by any label. For example most things in `Counterexamples`, `Archive`, in the root dir, `Mathilb/[Testing,Deprecated,Init]`. It seems unreasonable to me that the CI should fail on any PR modifying these labels.

However, I did add a test ensuring that all subfolders of `Mathlib/` are covered by a label, which might address your concerns in a way?

grunweg replied:
Good points - makes sense to not make CI fail. The current test is good enough for me!

grunweg reviewed (commented):
A few typos, and one last comment. After that, I'd maintainer merge (though perhaps @adomani) should rather do so).

grunweg commented:
Did you mean to remove the exit code 3 again?

grunweg commented:
(Otherwise, if CI passes etc, feel free to maintainer merge on my behalf - perhaps after asking Damiano's opinion as well.)

bryangingechen commented on scripts/autolabel.lean:
```diff
@@ -0,0 +1,203 @@
+      "Mathlib" / "RingTheory",
+      "Mathlib" / "GroupTheory",
+      "Mathlib" / "RepresentationTheory",
+      "Mathlib" / "LinearAlgebras"] },
```
```suggestion
      "Mathlib" / "LinearAlgebra"] },
```

adomani replied:
How about adding a `pathExists` check for all the dirs?

joneugster replied:
Sorry for the confusion, I was playing around, forgot to add the "WIP" label and went offline...

There is a check in place that all paths exist and the workflow fails otherwise!

adomani commented on scripts/autolabel.lean:
```diff
@@ -0,0 +1,256 @@
+
+- it ensures all paths in specified in `AutoLabel.mathlibLabels` exist
+- It makes sure all subfolders of `Mathlib/` belong to at least one label.
+  there is `AutoLabel.mathlibUnlabelled` to add exceptions for this test.
```
```suggestion
  There is `AutoLabel.mathlibUnlabelled` to add exceptions for this test.
```

bryangingechen commented on .github/workflows/add_label_from_diff.yaml:
```diff
@@ -0,0 +1,39 @@
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          GH_REPO: ${{ github.repository }}
+          NUMBER: ${{ github.event.number }}
```
Will this work even if the workflow is triggered with `workflow_dispatch` right?

joneugster replied:
I think I cannot test it since the dispatch button does not appear... or at least I do not get a dispatch button?

Maybe the dispatch button only appears for workflows on `master`?

joneugster replied:
Looking at google, I think it will not work... Maybe I remove the `workflow_dispatch` then...

bryangingechen replied:
Yeah, workflow_dispatch only works when it's in `master`

bryangingechen commented on lakefile.lean:
```diff
@@ -80,6 +80,13 @@ lean_lib docs where
 ## Executables provided by Mathlib
 -/
 
+/-- `lake exe autolabel 150100` adds a topic label to PR `150100` if there is a unique choice.
```
It would be good to mention that `gh` is requried to be installed for this to be work since even people contributing to mathlib might not have it / be aware of it (it's not mentioned in our info for contributors, right?)

Maybe it should also be mentioned in the module doc / docstrings in the lean file as well? (and in an error message if gh is missing?)

grunweg replied:
Doing both sounds good to me. (I haven't checked if `gh` needs to be configured some way, perhaps that can be added as well.)

joneugster replied:
Done.

* I did not add an extra error message if `gh` is not installed, but the user will get something like ``unknown command: `gh`.`` (or whatever `bash` says). And I added the note about requiring `gh` to every docstring the user could possibly look at.
* Installing `gh` is just following the official installation instructions (e.g. https://github.com/cli/cli?tab=readme-ov-file#installation) which are basically installing the os-dependent binaries and running `gh auth login` to authenticate, but no special setup.

bryangingechen commented:
As it seems Michael and Damiano are happy with this, in addition to my comment about `gh` above, please also fill out the commit message a little bit more (at least `lake exe autolabel` should be mentioned), and then we can test this out in production:
bors d+

adomani commented on scripts/autolabel.lean:
```diff
@@ -0,0 +1,286 @@
+      "Mathlib" / "Topology",
+      "Mathlib" / "AlgebraicTopology"] },
+  { label := "CI",
+    dirs := #[".github" / "workflows"] },
```
```suggestion
    dirs := #[".github"] },
```
the "main build" file is now in `.github` (although modifying it implies also modifying files inside `workflows`).

We catch a couple of other files in `.github` but not in `.github/workflows` as well.

adomani commented on scripts/autolabel.lean:
```diff
@@ -0,0 +1,286 @@
+-/
+unsafe def main (args : List String): IO Unit := do
+  if args.length > 1 then
+    println s!"autolabel: invalid number of arguments ({args.length}). Please run without \
+    arguments or provide the target PR's number as single argument!"
```
```suggestion
    println s!"autolabel: invalid number of arguments ({args.length}), expected at most 1. \
    Please run without arguments or provide the target PR's number as single argument!"
```

adomani commented on scripts/autolabel.lean:
```diff
@@ -0,0 +1,286 @@
+  for label in mathlibLabels do
+    for dir in label.dirs do
+      unless ← FilePath.pathExists dir do
+        println s!"::error file=scripts/autolabel.lean,line=60::directory {dir} does not exist! \
```
```suggestion
        println s!"::error file=scripts/autolabel.lean,line=60::directory {dir} does not exist! \
```
what does `line=60` refer to?

joneugster replied:
`line=84` (updated) is supposed to be roughly the definition of `def mathlibLabels : Array Label`?

I don't know a way to find this automatically. Should I just do `line=1` instead?

adomani replied:
This may work (untested):
```lean
 line := ((declRangeExt.find? (← getEnv) `Autolabel.mathlibLabels).getD default).range.pos.line
```

joneugster replied:
hmm, but I'm in `IO` and don't really have access to the `env`. Usually I would use `withImportModules` in this case, but that does not work since `script/autolabel.lean` is not importable.

grunweg replied:
How about saying "directory `foo` in `mathlibLabels` does not exist"? This avoids line numbers at all, but is hopefully equally informative.

joneugster replied:
done

adomani reviewed (approved):
I'm really happy with this version!

I also had autoparameters in one of the iterations of the EnvExtension PR.  I think that they worked less well there, but they look very good here!

Thanks!

bryangingechen commented on scripts/autolabel.lean:
```diff
@@ -0,0 +1,286 @@
+  -- test: validate that the labels cover all of the `Mathlib/` folder
+  let notMatchedPaths ← findUncoveredPaths "Mathlib" (exceptions := mathlibUnlabelled)
+  if notMatchedPaths.size > 0 then
+    -- note: only emmitting a warning because the workflow is only triggered on the first commit
```
```suggestion
    -- note: only emitting a warning because the workflow is only triggered on the first commit
```

bryangingechen commented on scripts/autolabel.lean:
```diff
@@ -0,0 +1,286 @@
+unsafe def main (args : List String): IO Unit := do
+  if args.length > 1 then
+    println s!"autolabel: invalid number of arguments ({args.length}), expected at most 1. \
+    Please run without arguments or provide the target PR's number as single argument!"
```
```suggestion
    Please run without arguments or provide the target PR's number as a single argument!"
```

joneugster commented:
Here are 4 test-PRs showing the expected outcomes:

- [ ] : [warning if topic labels do not cover everything](https://github.com/leanprover-community/mathlib4/pull/17166/files#diff-914f28d3d564a9e54089bda43485720d093959a94c038587b1edff9d0713268d)
- [ ] : [error if included path does not exist](https://github.com/leanprover-community/mathlib4/pull/17163/files#diff-914f28d3d564a9e54089bda43485720d093959a94c038587b1edff9d0713268d)
- [ ] : [error if excluded path does not exist](https://github.com/leanprover-community/mathlib4/pull/17167/files#diff-914f28d3d564a9e54089bda43485720d093959a94c038587b1edff9d0713268d)
- [ ] : [successfully added "CI" label](https://github.com/leanprover-community/mathlib4/pull/17168)

@jcommelin you wanted to look over this again and merge.

Since delegation, I modified the warning/error message (i.e. added `githubAnnotation`) and added `"Mathlib" / "Std"` as exception, since this seemed to have appeared in the last day or so.

jcommelin commented on scripts/autolabel.lean:
```diff
@@ -0,0 +1,308 @@
+        -- print github annotation error
+        println <| AutoLabel.githubAnnotation "error" "scripts/autolabel.lean"
+          s!"Misformatted `{ ``AutoLabel.mathlibLabels }`"
+          s!"directory {dir} does not exist! (from label {label.label})"
```
Should this error give a bit more detailed instructions to the unlucky user whose PR has a red cross mark?

joneugster replied:
Updated it to


> Misformatted `AutoLabel.mathlibLabels`
> 
> scripts/autolabel.lean: directory 'Mathlib/LinearAlgebras' does not exist but is included by label 't-algebra'. Please update `AutoLabel.mathlibLabels`!

(see examples)

Also I realised that I can add the following to the workflow:

```
on:
  push:
    paths:
      - scripts/autolabel.lean
      - .github/workflows/add_label_from_diff.yaml
```

Now the workflow will re-trigger whenever you modify the `autolabel` script, immediately throwing the error.

I think now the only way to trigger an error on a future PR is to merge a PR which deletes an entire subfolder of `Mathlib` but didn't do so in the commit present when the PR was created.
